### PR TITLE
expression: fix date format identifies '\n' as invalid separator

### LINKED ISSUE #32232: date_format with "\n" as pattern has different behavior with MySQL
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
create table t(dt datetime);

explain select * from t where dt <= date_format('2022-02-09', '%Y-%m-%d 16:33:00');

explain select * from t where dt <= date_format('2022-02-09', '%Y-%m-%d
16:33:00');
```

### 2. What did you expect to see? (Required)
The selection condition of execution plan in both sqls are `le(test.t.dt, 2022-02-09 16:33:00.000000)`

### 3. What did you see instead (Required)
The second is `le(test.t.dt, NULL)`

### 4. What is your TiDB version? (Required)
4.0, 5.0, 5.1, and older 5.1 versions.

### PR DESCRIPTION
**Pull request by mengxin9014**
### What problem does this PR solve?
fix date formate identifies '\n' is invalid separator

Issue Number: close #32232

Problem Summary:

### What is changed and how it works?
builtin_time.go

### Check List

Tests 
- [ ] Unit test
### Release note

```release-note
fix date formate identifies '\n' as invalid separator
```

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -786,7 +786,8 @@ func isValidSeparator(c byte, prevParts int) bool {
 		return true
 	}
 
-	if prevParts == 2 && (c == ' ' || c == 'T') {
+	// add c == '\n' is for https://github.com/pingcap/tidb/issues/32232
+	if prevParts == 2 && (c == ' ' || c == 'T' || c == '\n') {
```
Comment (windtalker) [inline types/time.go:790]: According to MySQL [code](https://github.com/mysql/mysql-server/blob/8.0/mysys/my_time.cc#L121), it allow all the `space` char.

Comment (mengxin9014) [reply types/time.go]: done

**Review by windtalker — APPROVED** [review]
LGTM

**Comment by LittleFall** [comment]
issue link of pr description is wrong

File: types/time.go [hunk 2]
```diff
@@ -786,7 +786,8 @@ func isValidSeparator(c byte, prevParts int) bool {
 		return true
 	}
 
-	if prevParts == 2 && (c == ' ' || c == 'T') {
+	// for https://github.com/pingcap/tidb/issues/32232
+	if prevParts == 2 && (c == 'T' || c == ' ' || c == '\t' || c == '\n' || c == '\v' || c == '\f' || c == '\r') {
```
Comment (LittleFall) [inline types/time.go:790]: consider use `Unicode.IsSpace`?

**Review by LittleFall — APPROVED** [review]

**Comment by mengxin9014** [comment]
#32232

**Pull request merged**